consensus/propagation/have_wants.go

```diff
@@ -63,58 +66,143 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
 		return
 	}
 
-	reqLimit := 1
-
-	// if enough requests have been made for the parts, don't request them.
-	for _, partIndex := range hc.GetTrueIndices() {
-		reqs := blockProp.countRequests(height, round, partIndex)
-		if len(reqs) >= reqLimit {
-			hc.SetIndex(partIndex, false)
-			// mark the part as fully requested.
-			fullReqs.SetIndex(partIndex, true)
+	for _, index := range hc.GetTrueIndices() {
+		select {
+		case <-blockProp.ctx.Done():
+		case p.requestChan <- request{height: height, round: round, index: uint32(index)}:
 		}
-		// don't request the part from this peer if we've already requested it
-		// from them.
-		for _, p := range reqs {
-			// p == peer means we have already requested the part from this peer.
-			if p == peer {
-				hc.SetIndex(partIndex, false)
+	}
+
+	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
+}
+
+// perPeerPerBlockConcurrentRequestLimit returns the maximum number of requests for a given block by peer.
+// it helps avoid downloading the whole block from a single peer.
+func perPeerPerBlockConcurrentRequestLimit() int64 {
+	// TODO implement
+	return 300
+}
+
+func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
+	var requestsBA *bits.BitArray
+	height := int64(0)
+	round := int32(0)
+	periodicRequestsDelay := 10 * time.Millisecond
+	periodicRequestTicker := time.NewTicker(periodicRequestsDelay)
+	for {
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case <-periodicRequestTicker.C:
+			// periodically send the existing requests
+			if requestsBA != nil && !requestsBA.IsEmpty() {
+				blockProp.sendWant(ps, height, round, requestsBA)
+				requestsBA = nil
+			}
```

Comment:
is this just randomly sending wants for every peer that we connect to?

side note: In order to figure out what is sent here, we have to read the entire function and keep track of when it changes in our heads at each step, which imo is implicit and arguably convoluted.

Reply:
This is part of the batching mechanism. This ensures sending periodically the wants we need to send even if we don't have a batch of a certain number of wants